### PR TITLE
Concurrency improvements for workers

### PR DESCRIPTION
**Pull request by Kixiron**
Depends on #900

* Removed barriers from within the worker runtime, got a 20% speedup
* Distribute update records even across workers, got a 15% speedup (compared to the barriers change, not master), scaling across threads now works much better

This PR is best reviewed per-commit

### CODE REVIEW

**Comment by ryzhyk** [comment]
20% + 80%, wow!  With how many workers?

**Comment by ryzhyk** [comment]
This looks super-linear going from 1 to 2 threads. How's this possible?

> ```
> citations/1 thread/100000 records
> time:   [23.410 s 23.874 s 24.336 s]
> change: [-9.7934% -8.0272% -6.1577%] (p = 0.00 < 0.05)
> 
> citations/2 threads/100000 records
> time:   [8.5376 s 8.6641 s 8.7864 s]
> change: [-63.134% -61.697% -60.309%] (p = 0.00 < 0.05)
> 
```

**Comment by Kixiron** [comment]
Ah, it seems I forgot to transfer over a rather crucial piece of code, this still has
```rust
    .filter(|&(relid, _)| self.is_leader() && self.program.get_relation(relid).input)
```
Which is rather disappointing, I knew something *had* to be off with the times but I really didn't want it to be

**Comment by Kixiron** [comment]
Ok, now I've gotten everything fixed up and we get steady (and realistic) improvements with more worker threads

*Percent change is relative to 0c054e1*
```
citations/1 thread/100000 records
time:   [23.434 s 23.570 s 23.687 s]
change: [-4.642% -5.496% -7.069%] (p = 0.00 < 0.05)

citations/2 threads/100000 records
time:   [18.210 s 18.699 s 19.186 s]
change: [-15.329% -15.216% -15.064%] (p = 0.00 < 0.05)

citations/3 threads/100000 records
time:   [18.973 s 19.525 s 19.983 s]
change: [-16.196% -15.090% -14.419%] (p = 0.00 < 0.05)

citations/4 threads/100000 records
time:   [23.698 s 24.439 s 25.188 s]
change: [+6.259% +7.490% +8.784%] (p = 0.00 < 0.05)
```

**Comment by ryzhyk** [comment]
Sorry about being slow to review this.  Will try to get around to it in a few days.

**Comment by Kixiron** [comment]
I might have made a rather significant breakthrough
```
twitter/4 threads/50000 records
time:   [41.328 ms 41.655 ms 42.008 ms]

twitter/3 threads/50000 records
time:   [43.092 ms 43.318 ms 43.571 ms]

twitter/2 threads/50000 records
time:   [60.999 ms 61.160 ms 61.330 ms]

twitter/1 thread/50000 records
time:   [97.935 ms 98.150 ms 98.345 ms]


twitter/4 threads/100000 records
time:   [83.116 ms 83.994 ms 84.843 ms]

twitter/3 threads/100000 records
time:   [86.612 ms 87.189 ms 87.937 ms]

twitter/2 threads/100000 records
time:   [123.47 ms 124.00 ms 124.41 ms]

twitter/1 thread/100000 records
time:   [200.01 ms 200.45 ms 200.89 ms]


twitter/4 threads/150000 records
time:   [130.07 ms 131.84 ms 133.62 ms]

twitter/3 threads/150000 records
time:   [131.03 ms 132.38 ms 133.87 ms]

twitter/2 threads/150000 records
time:   [189.16 ms 190.68 ms 192.97 ms]

twitter/1 thread/150000 records
time:   [302.63 ms 303.66 ms 304.60 ms]


twitter/4 threads/200000 records
time:   [172.05 ms 173.83 ms 175.85 ms]

twitter/3 threads/200000 records
time:   [179.01 ms 180.69 ms 182.53 ms]

twitter/2 threads/200000 records
time:   [252.50 ms 253.57 ms 254.60 ms]

twitter/1 thread/200000 records
time:   [408.65 ms 409.31 ms 410.01 ms]


twitter-macro/4 threads/10000000 records
time:   [16.766 s 17.046 s 17.308 s]

twitter-macro/3 threads/10000000 records
time:   [15.120 s 15.377 s 15.602 s]

twitter-macro/2 threads/10000000 records
time:   [19.230 s 19.589 s 19.962 s]

twitter-macro/1 thread/10000000 records
time:   [25.492 s 25.611 s 25.763 s]
```

**Comment by ryzhyk** [comment]
This is really cool. What changes made the most difference?

It looks like for some reason gitlab doesn't run your pipelines anymore. I might be getting confused by forced pushes that change history.  Let me try cloning this into my branch and running the pipeline there.

**Comment by ryzhyk** [comment]
There are some failures: https://gitlab.com/ddlog/differential-datalog-ryzhyk/-/pipelines/246255581

**Comment by Kixiron** [comment]
The failures are because I forgot a `max()`, the second to last commit was the largest difference, I guess because of batching commands but I'm not entirely sure to be honest

**Comment by ryzhyk** [comment]
All but one test pass now: https://gitlab.com/ddlog/differential-datalog-ryzhyk/-/pipelines/246707379

The stream test is timing out in CI, but works just fine on my machine.

**Comment by Kixiron** [comment]
It works for me locally too, I'm confused as to why CI is broken

**Comment by ryzhyk** [comment]
Hard-to-reproduce bugs are fun.  No worries, I'll try to figure it out when I finally get to reviewing this PR (hopefully just a few more days). In the meanwhile I'd love to merge the benchmarking PR.  I think the only two pressing issues there were moving datasets out of the repo and some way to run a scaled down version in CI (if it's not part of the test suite, it's guaranteed to bit rot).  Just wondering if you'd have a chance to look into that at some point.

**Comment by Kixiron** [comment]
I did the former, the datasets are now downloaded on-demand from SNAP. As for the latter, I can try to but I only have experience with github ci and not gitlab

**Comment by ryzhyk** [comment]
Oh, don't worry about actually adding new tests to the CI, can do that. I was just hoping you could create a scaled-down version of the benchmark that runs in 10 minutes or less, or perhaps there's a way to run the existing test this way.  Basically, I want to use the CI to make sure that benchmarks compile and work on each commit.  We can do the actual benchmarking offline for the time being.

**Comment by Kixiron** [comment]
Running only the twitter micro benches on either 50k or 100k records should be perfectly good for CI I think, later on once [iai](https://bheisler.github.io/criterion.rs/book/iai/iai.html) stabilizes I may integrate that so we can have consistent benchmarks on CI (iai uses callgrind to count instructions, so CI power fluctuations don't affect its results)

**Comment by ryzhyk** [comment]
Sorry again about the delay.  I am finally starting to look into this today.  My goal is to review the changes and try to reproduce/debug the failure that has so far only shown up in CI.  Is there a reason this is still a draft PR?

**Review by ryzhyk — COMMENTED** [review]
@Kixiron , here are my thoughts about this PR. Please let me know if they make sense.  I would like to try to fix it myself just to be sure I've fully internalized the code and haven't missed anything.

I think this design can be made to work. My old design relied on worker-to-worker communication for progress management with worker 0 being the only one to communicate directly with client threads.  The new implementation is moving towards a symmetric architecture with all workers receiving commands and sending acknowledgement.  However I see some leftovers of the old design in handling of `Flush` and `Stop`, which get broadcast to everyone, but only worker 0 replies to them. This leads to a bunch of unnecessary code (e.g., the `running` flag and `unpark_peers()`).  In addition, since the barrier has been removed in this PR, it is possible that worker 0 sends the acknowledgement before other workers have finished processing updates.

All we need to do complete this PR is have all workers reply to `Stop` and `Flush` and eliminate all worker-to-worker communication.  There are also a couple of minor things, like dropping the sessions for non-input relations (this tell DD that it shouldn't expect any updates from these sessions) and using `step_or_park` instead of `step` to prevent busy-waiting.

PS. I was able to reproduce the bug with the `stream` test getting stuck in CI.  I just had to change `-w 4` to `-w 16` in `test-stream.sh`. I can see that all worker threads are stuck in an infinite loop calling `step`.  Will try to figure our precisely what's going on tomorrow.

**Comment by Kixiron** [comment]
Yes, I totally agree. A lot of this is due to me trying to "reconcile" the two designs to minimize code churn, which is a fairly stupid thing to do in retrospect, but the nice thing is that now I can go through an give it a nice refactor!

**Comment by ryzhyk** [comment]
> Yes, I totally agree. A lot of this is due to me trying to "reconcile" the two designs to minimize code churn, which is a fairly stupid thing to do in retrospect, but the nice thing is that now I can go through an give it a nice refactor!

I think I'm nearly done.  A bunch more code is gone now and I find it much easier to reason about correctness of the design.

**Comment by ryzhyk** [comment]
PS. I think your last rebase did not go well.  Can you please rollback to the pre-rebase version, so I can submit a PR against your PR.

File: rust/template/differential_datalog/src/program/worker.rs [hunk 1]
```diff
@@ -158,182 +141,136 @@ impl<'a> DDlogWorker<'a> {
...
+                    // before flushing & compacting all previous timestamp's traces
+                    Msg::Flush { advance_to } => {
+                        self.advance(&mut sessions, &mut traces, advance_to);
                         self.flush(&mut sessions, &probe);
 
-                        //println!("flushed");
-                        self.reply_sender
-                            .send(Reply::FlushAck)
-                            .map_err(|e| format!("failed to send ACK: {}", e))?;
+                        // Only the leader sends back a flush acknowledgement even though
```
[the remark this answers is not in the document]
Comment (Kixiron) [reply rust/template/differential_datalog/src/program/worker.rs:167]: Yes, I agree, and I think that we should follow something like materialize's [`PendingPeek` system](https://github.com/MaterializeInc/materialize/blob/main/src/dataflow/src/server.rs#L793-L848) for answering queries

**Comment by ryzhyk** [comment]
https://github.com/Kixiron/differential-datalog/pull/1

**Comment by ryzhyk** [comment]
Closing.  Will create a fresh PR instead.

**Pull request closed without merging**